graphix/transpiler.py

```diff
@@ -941,6 +981,283 @@ def _rzz_command_opt(self, control_node, target_node, ancilla, angle):
+                [
+                    ancilla[0],
+                    ancilla[9],
+                    ancilla[11],
+                    ancilla[13],
+                    ancilla[17],
+                    ancilla[21],
+                    ancilla[26],
+                ],
+            ]
+        )
+        seq.append(
+            [
+                "Z",
+                control_node2,
+                [
+                    ancilla[1],
+                    ancilla[5],
+                    ancilla[7],
+                    ancilla[9],
+                    ancilla[11],
+                    ancilla[15],
+                    ancilla[21],
+                    ancilla[26],
+                ],
+            ]
+        )
+        seq.append(
+            [
+                "Z",
+                target_node,
+                [
+                    ancilla[2],
+                    ancilla[4],
+                    ancilla[6],
+                    ancilla[8],
+                    ancilla[10],
+                    ancilla[12],
+                    ancilla[16],
+                    ancilla[18],
+                    ancilla[22],
+                ],
+            ]
+        )
+        return control_node1, control_node2, target_node, seq
+
+    @classmethod
+    def _ccx_command_opt(self, control_node1, control_node2, target_node, ancilla):
+        """Optimized MBQC commands for CCX gate
+
+        Parameters
+        ---------
+        input_node : int
+            input node index
+        ancilla : int
+            ancilla node index
+
+        Returns
+        ---------
+        out_node_control : int
+            control node on graph after the gate
+        out_node_target : int
+            target node on graph after the gate
+        commands : list
+            list of MBQC commands
+        """
+        assert len(ancilla) == 6
+        seq = [["N", ancilla[0]], ["N", ancilla[1]]]  # assign new qubit labels
+        seq.extend([["N", ancilla[2]], ["N", ancilla[3]]])
+        seq.extend([["N", ancilla[4]], ["N", ancilla[5]]])
+        seq.append(["E", (ancilla[0], ancilla[3])])
+        seq.append(["E", (ancilla[0], control_node2)])
+        seq.append(["E", (ancilla[1], control_node2)])
+        seq.append(["E", (ancilla[2], control_node2)])
+        seq.append(["E", (ancilla[3], ancilla[5])])
+        seq.append(["E", (ancilla[3], control_node1)])
+        seq.append(["E", (ancilla[3], ancilla[4])])
+        seq.append(["E", (ancilla[4], control_node2)])
+        seq.append(["E", (control_node1, control_node2)])
+        seq.append(["M", ancilla[2], "XY", 0.25, [], [], 6])
+        seq.append(["M", ancilla[3], "XY", -0.25, [], [], 0])
+        seq.append(["M", ancilla[4], "XY", 0.25, [], [], 6])
+        seq.append(["M", ancilla[1], "XY", -0.25, [], [], 6])
+        seq.append(["M", ancilla[0], "XY", -0.25, [], [], 6])
+        seq.append(["M", ancilla[5], "XY", 0.25, [], [], 6])
+        # seq.append(["C", control_node1, 6]) # TODO: does not work?
```

Comment:
@shinich1 
After running `pattern.perform_pauli_measurements()`, I got a series of commands, and there is a command which includes `"C"`.
Running the following code raises `TypeError: CommandNode.__init__() got an unexpected keyword argument 'vop'` and I am not sure how to add `"C"` command in the measurement pattern.
```python
from graphix.transpiler import Circuit
import numpy as np

circuit = Circuit(3)
# prepare |000>
circuit.h(0)
circuit.h(1)
circuit.h(2)
# prepare |111>
circuit.x(0)
circuit.x(1)
circuit.x(2)

circuit.ccx(0, 1, 2)
pattern = circuit.transpile(opt = True)

pattern.standardize()
pattern.shift_signals()
pattern.perform_pauli_measurements()
pattern.minimize_space()

from qiskit.quantum_info import Statevector
ket = pattern.simulate_pattern(backend='statevector').flatten()
ket = Statevector(ket)
display(ket.draw(output='latex'))
```
Traceback of the error:
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[1], line 17
     14 circuit.ccx(0, 1, 2)
     15 pattern = circuit.transpile(opt = True)
---> 17 pattern.standardize()
     18 pattern.shift_signals()
     19 pattern.perform_pauli_measurements()

File [~/graphix/graphix/pattern.py:246](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/home/kazuki_main/graphix/~/graphix/graphix/pattern.py:246), in Pattern.standardize(self, method)
    234 """Executes standardization of the pattern.
    235 'standard' pattern is one where commands are sorted in the order of
    236 'N', 'E', 'M' and then byproduct commands ('X' and 'Z').
   (...)
    243     defaults to 'local'
    244 """
    245 if method == "local":
--> 246     localpattern = self.get_local_pattern()
    247     localpattern.standardize()
    248     self.seq = localpattern.get_pattern().seq

File [~/graphix/graphix/pattern.py:229](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/home/kazuki_main/graphix/~/graphix/graphix/pattern.py:229), in Pattern.get_local_pattern(self)
    227     if index in self.output_nodes:
    228         node_prop[index]["output"] = True
--> 229     node = CommandNode(index, **node_prop[index])
    230     nodes[index] = node
    231 return LocalPattern(nodes, self.output_nodes, morder)

TypeError: CommandNode.__init__() got an unexpected keyword argument 'vop'
```

Reply:
Hi, I would like to know how to skip specific Pauli measurements. Is it sufficient to simply add
```
if cmd[1] in nodes_skipped:
    continue
```
in the beginning of the following loop?

https://github.com/TeamGraphix/graphix/blob/de8692251d45947e42c2b57ae580d387c83b254b/graphix/pattern.py#L1633-L1649

Reply:
@king-p3nguin  Hi - yes, I think that should work.

Reply:
Hmm... I tried this method, but it is not working...
I skipped the 29th node Pauli measurement and got the following commands for opt, and implemented them as so, but the outcome is not correct.
```
N, node = 0
N, node = 1
N, node = 5
N, node = 9
N, node = 13
N, node = 21
N, node = 25
N, node = 28
N, node = 29
N, node = 30
E, nodes = (0, 25)
E, nodes = (1, 9)
E, nodes = (1, 25)
E, nodes = (5, 9)
E, nodes = (5, 25)
E, nodes = (9, 21)
E, nodes = (9, 29)
E, nodes = (13, 25)
E, nodes = (25, 29)
E, nodes = (29, 30)
M, node = 5, plane = XY, angle(pi) = 0.25, s-domain = [4, 2], t_domain = [], Clifford index = 6
M, node = 9, plane = XY, angle(pi) = -0.25, s-domain = [8, 6, 4, 2], t_domain = [], Clifford index = 0
M, node = 1, plane = XY, angle(pi) = -0.25, s-domain = [], t_domain = [], Clifford index = 6
M, node = 13, plane = XY, angle(pi) = 0.25, s-domain = [12, 10, 8, 6, 4, 2], t_domain = [], Clifford index = 6
M, node = 0, plane = XY, angle(pi) = -0.25, s-domain = [], t_domain = [], Clifford index = 6
M, node = 21, plane = XY, angle(pi) = 0.25, s-domain = [20, 14], t_domain = [], Clifford index = 6
M, node = 29, plane = XY, angle(pi) = 0, s-domain = [], t_domain = [], Clifford index = 0
X byproduct, node = 25, domain = [24]
X byproduct, node = 30, domain = [14, 20, 24, 26, 29]
X byproduct, node = 28, domain = [3, 5, 7, 9, 11, 13, 17, 19, 23]
Z byproduct, node = 25, domain = [0, 9, 11, 13, 17, 21, 27]
Z byproduct, node = 30, domain = [1, 5, 7, 9, 11, 15, 21, 27]
Z byproduct, node = 28, domain = [2, 4, 6, 8, 10, 12, 16, 18, 22]
```

Reply:
hmm, in such a case, I recommend to leave input qubits unmeasured as well. how would that go?

Reply:
I think there is still some bug in the transpiler. The following code gave $\ket{010}$, which is incorrect.
```
from graphix.transpiler import Circuit
import numpy as np
circuit = Circuit(3)
# prepare |000>
circuit.h(0)
circuit.h(1)
circuit.h(2)
# prepare |011>
circuit.x(1)
circuit.x(2)
# toffoli
circuit.h(2)
circuit.cnot(1, 2)
circuit.rz(2, -np.pi/4)
circuit.cnot(0, 2)
circuit.rz(2, np.pi/4)
circuit.cnot(1, 2)
circuit.rz(1, np.pi/4)
circuit.rz(2, -np.pi/4)
circuit.cnot(0, 2)
circuit.cnot(0, 1)
circuit.rz(2, np.pi/4)
circuit.rz(0, np.pi/4)
circuit.rz(1, -np.pi/4)
circuit.h(2)
circuit.cnot(0, 1)
# transpile
pattern = circuit.transpile()
pattern.standardize()
pattern.shift_signals()
pattern.perform_pauli_measurements()
pattern.minimize_space()
# draw with latex
from qiskit.quantum_info import Statevector
ket = pattern.simulate_pattern(backend='statevector').flatten()
ket = Statevector(ket)
ket.draw(output='latex')
```

Reply:
let me take a look

Reply:
Thanks for waiting, the bug is fixed, the patch is on [this branch ](https://github.com/TeamGraphix/graphix/tree/62-bug-pattern-optimization-result-in-wrong-output-state-in-some-cases) and will be merged very soon (#63).